src/coprocessor/trie/mod.rs

```diff
@@ -0,0 +1,734 @@
+pub type ChildMap<F, const ARITY: usize> = InversePoseidonCache<F>;
+
+/// A sparse Trie.
+pub struct Trie<'a, F: LurkField, const ARITY: usize, const HEIGHT: usize> {
+    root: F,
+    empty_roots: [F; HEIGHT],
+    hash_cache: &'a PoseidonCache<F>,
+    children: &'a mut ChildMap<F, ARITY>,
+    count: usize,
+}
+
+pub struct LookupProof<F: LurkField, const ARITY: usize, const HEIGHT: usize> {
+    preimage_path: PreimagePath<F, ARITY>,
+}
+
+impl<F: LurkField, const ARITY: usize, const HEIGHT: usize> LookupProof<F, ARITY, HEIGHT> {
+    fn new(preimage_path: PreimagePath<F, ARITY>) -> Self {
+        Self { preimage_path }
+    }
+
+    /// Verify a `LookupProof`. Note that this verification is exactly what must be proved in the circuit.
+    pub fn verify(&self, root: F, key: F, value: F, hash_cache: &PoseidonCache<F>) -> bool {
+        let path = Trie::<F, ARITY, HEIGHT>::path(key);
+
+        let mut next = root;
+        for (k, preimage) in path.iter().zip(&self.preimage_path) {
+            let computed_hash = Trie::<F, ARITY, HEIGHT>::compute_hash(hash_cache, *preimage);
+
+            if next != computed_hash {
+                return false;
+            }
+            next = preimage[*k];
+        }
+        next == value
+    }
+}
+
+pub struct InsertProof<F: LurkField, const ARITY: usize, const HEIGHT: usize> {
+    old_proof: LookupProof<F, ARITY, HEIGHT>,
+    new_proof: LookupProof<F, ARITY, HEIGHT>,
+}
+
+impl<F: LurkField, const ARITY: usize, const HEIGHT: usize> InsertProof<F, ARITY, HEIGHT> {
+    fn new(
+        old_proof: LookupProof<F, ARITY, HEIGHT>,
+        new_proof: LookupProof<F, ARITY, HEIGHT>,
+    ) -> Self {
+        Self {
+            old_proof,
+            new_proof,
+        }
+    }
+
+    /// Verify an `InsertProof`. Note that this verification is exactly what must be proved in the circuit.
+    pub fn verify(
+        &self,
+        old_root: F,
+        new_root: F,
+        key: F,
+        old_value: Option<F>,
+        new_value: F,
+        hash_cache: &PoseidonCache<F>,
+    ) -> bool {
+        let old_verified = self.old_proof.verify(
+            old_root,
+            key,
+            old_value.unwrap_or_else(|| F::zero()),
+            hash_cache,
+        );
+
+        let new_verified = self.new_proof.verify(new_root, key, new_value, hash_cache);
+
+        let paths_differ_by_at_most_one = self
+            .old_proof
+            .preimage_path
+            .iter()
+            .zip(&self.new_proof.preimage_path)
+            .all(|(a, b)| {
+                let differing_postition_count = a
+                    .iter()
+                    .zip(b)
+                    .map(|(x, y)| x != y)
+                    .fold(0, |acc, are_diff| acc + are_diff as usize);
+                differing_postition_count == 1 || differing_postition_count == 0
+            });
+
+        old_verified && new_verified && paths_differ_by_at_most_one
```

Comment:
Nit: `new_verified` is a bit expensive: can we check for equality with `old_proof` if paths differ by 0? We perhaps don't care and could add a TODO instead.

Reply:
Hmmm. Insofar as the vanilla implementation is intended to provide a demonstrative spec of what the circuit needs to verify, we shouldn't optimize it away. It will need to be proved in a flat circuit where this is impossible.

That said, the `paths_differ_by_at_most_one` calculation is also shortcutting and will need to be extrapolated/translated at circuit implementation time. What you suggest is theoretically okay, as a minor evaluation-time optimization — as long as it doesn't confuse circuit implementers. I'll play with whether I can do it in a way that's simple/clear enough.